imkar/scattering/analytical.py

```diff
@@ -0,0 +1,186 @@
+    In this solution the parameter r in equation (9) is also limited to
+    -N <= r <= N. The width of the well is described by the gap_width
+    parameter and replaces the parameter 2A.
+
+    """
+
+    # check inputs
+    frequencies = np.atleast_1d(np.asarray(frequencies, dtype=float))
+    if frequencies.ndim > 1:
+        raise ValueError("frequencies has to be 1-dimensional")
+
+    incident_angles = np.atleast_1d(np.asarray(incident_angles, dtype=float))
+    if incident_angles.ndim > 1:
+        raise ValueError("incident_angles has to be 1-dimensional")
+    if (np.any(incident_angles <= 0)) or (np.any(incident_angles >= 90)):
+        raise ValueError("incident_angles have to be between 0° and 90°")
+
+    if not isinstance(rectangle_width, numbers.Real) or rectangle_width <= 0:
+        raise TypeError("rectangle_width has to be a real number >0")
+    if not isinstance(gap_width, numbers.Real) or gap_width <= 0:
+        raise TypeError("gap_width has to be a real number >0")
+    if not isinstance(height, numbers.Real) or height <= 0:
+        raise TypeError("height has to be a real number >0")
+    if not isinstance(speed_of_sound, numbers.Real) or speed_of_sound <= 0:
+        raise TypeError("speed_of_sound has to be a real number >0")
+
+    # Initialization
+    incident_angles = incident_angles*np.pi/180  # from degree to radiant
+    # vector of corresponding wavelength in air
+    lambda_air = speed_of_sound/frequencies
+    k = 2*np.pi/lambda_air  # vector of wavenumbers
+    eps = 1E-3  # stop criterion for infinite sum over u
+    n_max = 50  # truncation parameter for numbers of outgoing waves
+    n_max2 = 2*n_max+1  # counter from -N to N including zero
+    length = gap_width + rectangle_width  # length of spatial period
+
+    x_u = np.zeros((1, n_max), dtype=complex)
+    n = np.arange(-n_max, n_max+1)[:, np.newaxis]
+    previous_frequency = 0
+
+    n_frequency = frequencies.size  # number of computed frequencies
+    n_phi = incident_angles.size  # number of computed incident angles
+    scattering_coefficient = np.zeros((n_frequency, n_phi))
+
+    # calculation
+    for iFrequencies in range(0, n_frequency):
+        for iPhi in range(0, n_phi):
+            phi_0 = incident_angles[iPhi]
+            alpha_n = np.cos(phi_0) + n*lambda_air[iFrequencies]/length
+            beta_n = np.conj(-(1j*np.emath.sqrt(alpha_n**2-1)))
+
+            # helping variables
+            k_frequency = k[iFrequencies]
+            k_alpha_n = (k_frequency*alpha_n[:]).T
+            jk_alpha_n_over_width = 1j*k_alpha_n/gap_width
+
+            y_nr = np.zeros((n_max2, n_max2), dtype=complex)
+
+            for iN in range(1, (n_max2+1)):
+                u_max = np.zeros((1, n_max2), dtype=int)
+                u_max_element = 0
+                for iR in range(1, (n_max2+1)):
+                    u_max[0, iR-1] = 0
+                    previous_temp_U = 0
+                    err = 1
+                    while err > eps:
```

Comment:
Could this lead to non-convergence or is it save? If not, we could think about a conditional exit condition or make the tolerance an addutional input parameter of the function.